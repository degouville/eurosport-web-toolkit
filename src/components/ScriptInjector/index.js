import React from 'react';
import PropTypes from 'prop-types';
import Helmet from 'react-helmet';

class ScriptInjector extends React.Component {
  componentDidMount() {
    const { isServer, src, innerHTML, injectPlace, async, id, onLoad } = this.props;

    if (isServer) return;

    if (id && document.getElementById(id) !== null) return;

    this.script = document.createElement('script');

    if (src) {
      this.script.src = src;
    }

    if (onLoad) {
      this.script.onload = () => onLoad();
    }

    if (async) {
      this.script.async = true;
    }

    if (id) {
      this.script.id = id;
    }

    this.script.innerHTML = innerHTML;

    document[injectPlace].appendChild(this.script);
  }

  componentWillUnmount() {
    const { injectPlace } = this.props;

    if (this.script) {
      document[injectPlace].removeChild(this.script);
    }
  }

  render() {
    const { id, src, async, isServer, innerHTML, injectPlace } = this.props;

    if (isServer) {
      const ssrProps = {
        src,
        async,
        id,
      };

      if (injectPlace === 'body') {
        return innerHTML ? (
          // eslint-disable-next-line react/no-danger
          <script {...ssrProps} dangerouslySetInnerHTML={{ __html: innerHTML }} />
        ) : (
          <script {...ssrProps} />
        );
      }

      // eslint-disable-next-line react/no-danger-with-children
      return <Helmet>{innerHTML ? <script {...ssrProps}>{`${innerHTML}`}</script> : <script {...ssrProps} />}</Helmet>;
    }
    return null;
  }
}

ScriptInjector.defaultProps = {
  src: '',
  innerHTML: '',
  async: false,
  injectPlace: 'head',
  id: undefined,
  onLoad: undefined,
};

ScriptInjector.propTypes = {
  isServer: PropTypes.bool.isRequired,
  id: PropTypes.string,
  src: PropTypes.string,
  innerHTML: PropTypes.string,
  injectPlace: PropTypes.oneOf(['head', 'body']),
  async: PropTypes.bool,
  onLoad: PropTypes.func,
};

export default ScriptInjector;
