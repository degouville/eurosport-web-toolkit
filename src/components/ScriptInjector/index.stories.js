import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, boolean } from '@storybook/addon-knobs';
import { ScriptInjector } from '../..';

const stories = storiesOf('Components|ScriptInjector', module);

const innerHTML = `
  (() => alert('hello world'))();
`;

stories.add(
  'ServerSide',
  () => (
    <ScriptInjector
      isServer
      src={text('src', 'url.to.custom.script.js')}
      async={boolean('async', true)}
      innerHTML={text('innerHTML', innerHTML)}
    />
  ),
  {
    info: 'injectPlace prop is not used for server side renderring',
  }
);

const clientSideScriptSrc = 'scriptInjector.client.side';
const injectPlaceMessage = 'injectPlace prop is used only client - side';

stories
  .add(
    'ClientSide - in body',
    () => <ScriptInjector isServer={false} src={`${clientSideScriptSrc}.in.body`} injectPlace="body" async={false} />,
    {
      info: injectPlaceMessage,
    }
  )
  .add(
    'ClientSide - async in head',
    () => <ScriptInjector isServer={false} async src={`${clientSideScriptSrc}.in.head`} />,
    {
      info: injectPlaceMessage,
    }
  )
  .add(
    'ClientSide - with inner script - in body',
    () => <ScriptInjector isServer={false} innerHTML={`alert('script injected in body')`} injectPlace="body" />,
    {
      info: injectPlaceMessage,
    }
  )
  .add(
    'ClientSide - with inner script - in head',
    () => <ScriptInjector isServer={false} innerHTML={`alert('script injected in head')`} />,
    {
      info: injectPlaceMessage,
    }
  );
