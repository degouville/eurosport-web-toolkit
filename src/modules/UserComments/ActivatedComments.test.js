import React from 'react';
import { shallow } from 'enzyme';
import ActivatedComments, { StyledContentCss, StyledSpinnerWrapper } from './ActivatedComments';
import activatedCommentsConfig from './mockData/activatedComments.json';
import ScriptInjector from '../../components/ScriptInjector';

const requireSpy = jest.fn();

class ActivatedCommentsMock {
  constructor(spy) {
    this.spy = spy;
  }

  require(opts) {
    this.spy(opts);
    return this;
  }
}

describe('ActivatedComments', () => {
  let wrapper;

  beforeEach(() => {
    global.Livefyre = new ActivatedCommentsMock(requireSpy);
    wrapper = shallow(
      <ActivatedComments
        livefyreConfig={activatedCommentsConfig}
        loginCallback={() => null}
        logoutCallback={() => null}
        userToken=""
      />
    );
  });

  it('renders an activated UserComments Component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the loader waiting for the script to be loaded', () => {
    expect(wrapper.find(StyledSpinnerWrapper)).toHaveLength(1);
    expect(wrapper.find(StyledContentCss)).toHaveLength(0);
  });

  it('renders the Livefyre widget once script is loaded', () => {
    wrapper.setState({ isWidgetLoaded: true });
    expect(wrapper.find(StyledContentCss)).toHaveLength(1);
    expect(wrapper.find(StyledSpinnerWrapper)).toHaveLength(0);
  });

  it('should pass Livefyre script src to ScriptInjector', () => {
    expect(wrapper.find(ScriptInjector).props().src).toBe(activatedCommentsConfig.config.scriptUrl);
  });

  describe('Livefyre init', () => {
    it('should init Livefyre Widget when script is loaded', () => {
      wrapper.find(ScriptInjector).simulate('load');
      expect(requireSpy).toHaveBeenCalledWith(['fyre.conv#3', 'auth']);
    });

    it('should get getNetworkConfig with the good values', () => {
      const networkConfig = ActivatedComments.getNetworkConfig(activatedCommentsConfig);
      expect(networkConfig).toEqual({
        network: activatedCommentsConfig.config.networkDomain,
        strings: activatedCommentsConfig.labels,
      });
    });

    it('should get getConvConfig with the good values', () => {
      const convConfig = ActivatedComments.getConvConfig(activatedCommentsConfig);
      expect(convConfig).toEqual({
        siteId: activatedCommentsConfig.config.siteId,
        articleId: activatedCommentsConfig.collection.articleId,
        el: 'livefyre',
        collectionMeta: activatedCommentsConfig.collection.collectionMeta,
        checksum: activatedCommentsConfig.collection.checksum,
      });
    });
  });
});
