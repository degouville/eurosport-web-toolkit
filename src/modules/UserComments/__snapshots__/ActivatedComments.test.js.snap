// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivatedComments renders an activated UserComments Component 1`] = `
<Fragment>
  <ScriptInjector
    async={false}
    injectPlace="head"
    innerHTML=""
    isServer={false}
    onLoad={[Function]}
    src="//cdn.livefyre.com/Livefyre.js"
  />
  <Styled(div)>
    Comments
  </Styled(div)>
  <Styled(div)>
    <Styled(div)
      id="livefyre"
    >
      <Styled(div)>
        <Styled(Spinner)
          className=""
          color="#35D0FE"
          width="80px"
        />
      </Styled(div)>
    </Styled(div)>
    <Styled(div) />
  </Styled(div)>
  <Styled(a)
    href="http://livefyre.fr"
  >
    Powered by Livefyre
  </Styled(a)>
</Fragment>
`;
